src/libraries/Common/src/System/Reflection/Metadata/TypeNameParserHelpers.cs

```diff
@@ -0,0 +1,747 @@
+                true, // U+006D 'm'
+                true, // U+006E 'n'
+                true, // U+006F 'o'
+                true, // U+0070 'p'
+                true, // U+0071 'q'
+                true, // U+0072 'r'
+                true, // U+0073 's'
+                true, // U+0074 't'
+                true, // U+0075 'u'
+                true, // U+0076 'v'
+                true, // U+0077 'w'
+                true, // U+0078 'x'
+                true, // U+0079 'y'
+                true, // U+007A 'z'
+                true, // U+007B '{'
+                true, // U+007C '|'
+                true, // U+007D '}'
+                true, // U+007E '~'
+                false, // U+007F (DEL)
+            ];
+        }
+
+        internal static int GetIndexOfFirstInvalidAssemblyNameCharacter(ReadOnlySpan<char> input, bool strictMode)
+            => GetIndexOfFirstInvalidCharacter(input, strictMode, assemblyName: true);
+
+        internal static int GetIndexOfFirstInvalidTypeNameCharacter(ReadOnlySpan<char> input, bool strictMode)
+            => GetIndexOfFirstInvalidCharacter(input, strictMode, assemblyName: false);
+
+        internal static ReadOnlySpan<char> GetName(ReadOnlySpan<char> fullName)
+        {
+#if NET8_0_OR_GREATER
+            int offset = fullName.LastIndexOfAny(_endOfTypeNameDelimitersSearchValues);
+            Debug.Assert(offset != 0, "The provided full name must be valid");
+
+            if (offset > 0 && fullName[offset - 1] == EscapeCharacter) // this should be very rare (IL Emit & pure IL)
+            {
+                offset = GetUnescapedOffset(fullName, startIndex: offset);
+            }
+#else
+            int offset = GetUnescapedOffset(fullName, startIndex: fullName.Length - 1);
+#endif
+            return offset < 0 ? fullName : fullName.Slice(offset + 1);
+
+            static int GetUnescapedOffset(ReadOnlySpan<char> fullName, int startIndex)
+            {
+                int offset = startIndex;
+                for (; offset >= 0; offset--)
+                {
+                    if (fullName[offset] is '.' or '+')
+                    {
+                        if (offset == 0 || fullName[offset - 1] != EscapeCharacter)
+                        {
+                            break;
+                        }
+                        offset--; // skip the escaping character
+                    }
+                }
+                return offset;
+            }
+        }
+
+        internal static string GetRankOrModifierStringRepresentation(int rankOrModifier)
+        {
+            return rankOrModifier switch
+            {
+                ByRef => "&",
+                Pointer => "*",
+                SZArray => "[]",
+                1 => "[*]",
+                _ => ArrayRankToString(rankOrModifier)
+            };
+
+            static string ArrayRankToString(int arrayRank)
+            {
+                Debug.Assert(arrayRank >= 2 && arrayRank <= 32);
```

Comment:
If the max arrayRank is 32, the else branch below needn't use ValueStringBuilder. It could just stackalloc the max, e.g.
```C#
Span<char> buffer = stackalloc char[33];
buffer[0] = '[';
for (int i = 1; i < arrayRank; i++)
{
    buffer[i] = ',';
}
buffer[arrayRank] = ']';
return buffer.Slice(0, arrayRank + 1).ToString();
```
Probably not worth special-casing .NET 8+ at that point.

Reply:
> Probably not worth special-casing .NET 8+ at that point.

Thanks for the suggestion. I am happy to remove the `#if` defs (I feel dirty after writing this code).